crates/dbs-virtio-devices/src/device.rs

```diff
@@ -0,0 +1,582 @@
+            .map(|shms| (shms.host_addr, shms.guest_addr.raw_value()))
+    }
+
+    /// Get a shared reference to the guest memory object.
+    pub fn lock_guest_memory(&self) -> AS::T {
+        self.vm_as.memory()
+    }
+}
+
+/// Shared Memory between device and guest
+#[derive(Clone, PartialEq, Versionize, Debug)]
+pub struct VirtioSharedMemory {
+    /// offset from the base
+    pub offset: u64,
+    /// len of this shared memory region
+    pub len: u64,
+}
+
+/// A list of Shared Memory regions
+#[derive(Clone, Debug)]
+pub struct VirtioSharedMemoryList {
+    /// Host address
+    pub host_addr: u64,
+    /// Guest address
+    pub guest_addr: GuestAddress,
+    /// Length
+    pub len: GuestUsize,
+    /// kvm_userspace_memory_region flags
+    pub kvm_userspace_memory_region_flags: u32,
+    /// kvm_userspace_memory_region slot
+    pub kvm_userspace_memory_region_slot: u32,
+    /// List of shared regions.
+    pub region_list: Vec<VirtioSharedMemory>,
+
+    /// List of mmap()ed regions managed through GuestRegionMmap instances. Using
+    /// GuestRegionMmap will perform the unmapping automatically when the instance
+    /// is dropped, which happens when the VirtioDevice gets dropped.
+    ///
+    /// GuestRegionMmap is used instead of MmapRegion. Because We need to insert
+    /// this region into vm_as，but vm_as uses GuestRegionMmap to manage regions.
+    /// If MmapRegion is used in here, the MmapRegion needs to be clone() to create
+    /// new GuestRegionMmap for vm_as. MmapRegion clone() will cause the problem of
+    /// duplicate unmap during automatic drop, so we should try to avoid the clone
+    /// of MmapRegion. This problem does not exist with GuestRegionMmap because
+    /// vm_as and VirtioSharedMemoryList can share GuestRegionMmap through Arc.
+    pub mmap_region: Arc<GuestRegionMmap>,
+}
+
+/// A callback for the VMM to insert memory region for virtio devices that
+/// has device memory, such as DAX of virtiofs, pmem.
+///
+/// insert_region function is used to solve the problem that the virtio device cannot
+/// find the host address corresponding to the guest address when reading the
+/// guest device memory.
+///
+/// For example, the guest application executes the following code:
+/// {
+///     // "dax_fd" is virtio-fs file that support dax
+///     // "no_dax_fd" is virtio-fs file that do not support dax
+///     void *dax_ptr = (void*)mmap(NUMM, 4096, PORT, MAP_SHARED, dax_fd, 0);
+///     write(no_dax_fd, dax_ptr, 4096);
+/// }
+/// dragonball will coredump.
+///
+/// This is because the virtiofs device cannot resolve the dax_ptr address
+/// when calling vm_as.get_slice(). There is no DAX region in vm_as. This
+/// trait inserts the virtio device memory region, such as DAX region, into
+/// vm_as. This trait should be implemented in VMM when creating virtio
+/// devices with device memory, because the virtio device does not have
+/// permission to change vm_as.
+pub trait VirtioRegionHandler: Send {
```

Comment:
I think we don't need this abstraction right now.

Reply:
The next pr handle it.